### PR TITLE
[PyTorch] Fix write-after-free (TSAN) in GraphTask::set_error()

### PR DESCRIPTION
**Pull request by jjlilley**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33156 [PyTorch] Fix write-after-free (TSAN) in GraphTask::set_error()**

When dist_autograd_spawn_thrift's 'test_backward_node_failure_python_udf' test is
run, it was encountering a TSAN error related to holding the mutex while the
underlying datastructure was being dealloced.

In this change, we simply get a shared_ptr<> reference to the future, and
set_exception() without having the lock held, to avoid deallocing underneath
the lock.

In real life, this could cause memory corruption, since it's technically a use-after-free.

Differential Revision: [D19821362](https://our.internmc.facebook.com/intern/diff/D19821362/)

### CODE REVIEW

**Comment by jjlilley** [comment]
Checking: Is there somebody else that I should add as a reviewer?
This change fixes an actual [small] memory stomp in autograd.

**Comment by albanD** [comment]
@pritamdamania87 will you have time to double check this?

**Comment by pritamdamania87** [comment]
@jjlilley Thanks for fixing this, although I'm not sure if I understood the cause of this failure completely. Which structure was being deallocated while we hold the lock?

**Comment by jjlilley** [comment]
The GraphTask is being deallocated while its lock is held.
Here's a paste of LOG(INFO) logs and a trace that repros: P126170360

IIRC (recompiling to remember from earlier), the problem was that the task was actually owned by the DistAutogradContext.

Notice the calls to cleanupBackwardPass() in dist_engine.cpp, when the future is satisfied with an error:
   DistEngine::getInstance().cleanupBackwardPass(autogradContext);

This will in turn call autogradContext->resetGraphTask() and clear out the graphTask.

Since the shared_ptr<> params were shared_ptr<>& references, the pointer disappears from underneath the active calls.

And then when we go to unlock the mutex, the struct holding the mutex has been dealloced. This will stomp the freed memory (probably just an int32_t?), hopefully it's not being used again already.

fwiw, one slightly simpler alternative would be to make thread_on_exception() take a shared_ptr<> rather than const shared_ptr<>& - which extend the scope of the shared_ptr<>. That said, it's usually not super ideal to satisfy a future with a lock held, usually a recipe for deadlocks as the callbacks are triggered.

**Comment by jjlilley** [comment]
Could somebody please take a look? This is [minorly] corrupting memory.

File: torch/csrc/autograd/engine.cpp [hunk 1]
```diff
@@ -386,16 +386,18 @@ void Engine::thread_on_exception(
 void GraphTask::set_exception(
     std::exception& e,
     const std::shared_ptr<Node>& fn) {
-  std::lock_guard<std::mutex> lock(mutex_);
+  std::unique_lock<std::mutex> lock(mutex_);
   if (!has_error_.load()) {
     if (AnomalyMode::is_enabled() && fn) {
       fn->metadata()->print_stack();
```
Comment (pritamdamania87) [inline torch/csrc/autograd/engine.cpp]: Thanks for the detailed explanation, although can't we have the same problem with this patch? The GraphTask can be deallocated here, before we unlock right?

Shouldn't the fix be something like what I've done in https://github.com/pytorch/pytorch/pull/33214, where we hold the lock before calling `set_exception` in the callback for `addOutstandingRpc`?

Comment (jjlilley) [reply torch/csrc/autograd/engine.cpp]: Not sure I see how GraphTask can be deallocated in print_stack()?  (I admit that I'm newish to this module)
The cases that I observed in particular involved the future callback triggering the dist context to clear its current task - which this avoids by dropping the lock prior to the potentially triggered delete.

btw, aside, re: #33214 that you referenced - can you possibly avoid changing the context's std::mutex to std::recursive_mutex?

**Review by pritamdamania87 — APPROVED** [review]
Thanks for fixing this! Lets just add a comment in terms of why we need to unlock there.

**Comment by jjlilley** [comment]
@pritamdamania87 Per our offline conversation, I added a comment to make the reason for the change more clear.

**Pull request closed without merging**